sdk/identity/identity/src/msal/browserFlows/msalAuthCode.ts

```diff
@@ -3,249 +3,352 @@
+      authority,
+      knownAuthorities: getKnownAuthorities(tenantId, authority, options.disableInstanceDiscovery),
+      // If the users picked redirect as their login style,
+      // but they didn't provide a redirectUri,
+      // we can try to use the current page we're in as a default value.
+      redirectUri: options.redirectUri || self.location.origin,
+    },
+    cache: {
       cacheLocation: "sessionStorage",
       storeAuthStateInCookie: true, // Set to true to improve the experience on IE11 and Edge.
-    };
-    this.msalConfig.system = {
+    },
+    system: {
       loggerOptions: {
-        loggerCallback: defaultLoggerCallback(this.logger, "Browser"),
+        loggerCallback: defaultLoggerCallback(options.logger, "Browser"),
         logLevel: getMSALLogLevel(getLogLevel()),
         piiLoggingEnabled: options.loggingOptions?.enableUnsafeSupportLogging,
       },
+    },
+  };
+}
+
+// We keep a copy of the redirect hash.
+const redirectHash = self.location.hash;
+
+/**
+ * Uses MSAL Browser 2.X for browser authentication,
+ * which uses the [Auth Code Flow](https://learn.microsoft.com/en-us/azure/active-directory/develop/v2-oauth2-auth-code-flow).
+ * @internal
+ */
+export function createMsalBrowserClient(options: MsalBrowserFlowOptions): MsalBrowserClient {
+  const loginStyle = options.loginStyle;
+  if (!options.clientId) {
+    throw new CredentialUnavailableError("A client ID is required in browsers");
+  }
+  const clientId = options.clientId;
+  const logger = options.logger;
+  const tenantId = resolveTenantId(logger, options.tenantId, options.clientId);
+  const additionallyAllowedTenantIds: string[] = resolveAdditionallyAllowedTenantIds(
+    options?.tokenCredentialOptions?.additionallyAllowedTenants,
+  );
+  const authorityHost = options.authorityHost;
+  const msalConfig = generateMsalBrowserConfiguration(options);
+  const disableAutomaticAuthentication = options.disableAutomaticAuthentication;
+  let app: msalBrowser.IPublicClientApplication;
+  const loginHint = options.loginHint;
+
+  let account: AuthenticationRecord | undefined;
+  if (options.authenticationRecord) {
+    account = {
+      ...options.authenticationRecord,
+      tenantId,
     };
-    if (options.authenticationRecord) {
-      this.account = {
-        ...options.authenticationRecord,
-        tenantId: this.tenantId,
-      };
-    }
   }
 
-  private async getApp(): Promise<msalBrowser.IPublicClientApplication> {
-    if (!this.app) {
+  /**
+   * Return the MSAL account if not set yet
+   * @returns MSAL application
+   */
+  async function getApp(): Promise<msalBrowser.IPublicClientApplication> {
+    if (!app) {
       // Prepare the MSAL application
-      this.app = await msalBrowser.PublicClientApplication.createPublicClientApplication(
-        this.msalConfig as msalBrowser.Configuration,
+      app = await msalBrowser.PublicClientApplication.createPublicClientApplication(
+        msalConfig as msalBrowser.Configuration,
```

Comment:
is the `as msalBrowser.Configuration` needed? Can we do anything to remove the cast? That'll help us ensure we're providing the right shape to MSAL if we need to make changes